src/iosource/PktSrc.cc

```diff
@@ -276,39 +279,43 @@ bool PktSrc::GetCurrentPacket(const Packet** pkt)
 
 double PktSrc::GetNextTimeout()
 	{
-	bool pkt_available = have_packet;
-
-	if ( props.selectable_fd == -1 || run_state::pseudo_realtime )
-		pkt_available = ExtractNextPacketInternal();
-
 	if ( run_state::is_processing_suspended() )
 		return -1;
 
-	// If there's no file descriptor for the source, which is the case for some interfaces like
-	// myricom, we can't rely on the polling mechanism to wait for data to be available. As gross
-	// as it is, just spin with a short timeout here so that it will continually poll the
-	// interface. The old IOSource code had a 20 microsecond timeout between calls to select()
-	// so just use that.
-	if ( props.selectable_fd == -1 )
+	// If we're in pseudo-realtime mode, find the next time that a packet is ready
+	// and have poll block until then.
+	if ( run_state::pseudo_realtime )
 		{
-		if ( ! pkt_available && ! run_state::pseudo_realtime )
-			return 0.00002;
+		ExtractNextPacketInternal();
+
+		// This duplicates the calculation used in run_state::check_pseudo_time().
+		double pseudo_time = current_packet.time - run_state::detail::first_timestamp;
+		double ct = (util::current_time(true) - run_state::detail::first_wallclock) *
+		            run_state::pseudo_realtime;
+		return std::max(0.0, pseudo_time - ct);
 		}
-	// If we're live we want poll to do what it has to with the file descriptor. If we're not live
-	// but we're not in pseudo-realtime mode, let the loop just spin as fast as it can. If we're
-	// in pseudo-realtime mode, find the next time that a packet is ready and have poll block until
-	// then.
-	else if ( IsLive() )
-		return -1;
 
-	if ( ! run_state::pseudo_realtime )
-		return 0;
+	// If there's no file descriptor for the source, which is the case for some interfaces
+	// like myricom, we can't rely on the polling mechanism to wait for data to be
+	// available. As gross as it is, just spin with a short timeout here so that it will
+	// continually poll the interface. The old IOSource code had a 20 microsecond timeout
+	// between calls to select() so just use that.
+
+	// A heuristic to avoid short sleeps when a non-selectable live packet source has
+	// more packets queued is to return 0.0 if the source has yielded a packet on the
+	// last call to ExtractNextPacket() or has a packet ready.
+	//
+	// A non-selectable offline packet source is assumed to always be ready.
+	if ( props.selectable_fd == -1 )
+		{
+		if ( have_packet || had_packet || ! IsLive() )
```

Comment:
With respect to the assumption about offline packet sources being always ready, I am wondering whether the `! IsLive()` check can be omitted. If the assumption holds, I would have expected `had_packet` to be true at this point anyway.

Another idea would be to make this more explicit by introducing a `IsReady/Idle()` method so that the packet source can influence the decision. The base class could provide the `had_packet` logic as a default. However, I might be overthinking this, as Packet Sources without FDs should probably just override `GetNextTimeout()` and replicate the suspended- and pseudo-realtime logic if necessary.

Reply:
> If the assumption holds, I would have expected had_packet to be true at this point anyway.

Good callout. At least for libpcap that should hold and I agree this adds/keeps a special case while one motivation was to streamline things. A non-selectable and non-ready offline packet source being subject to the same small sleep actually sounds more reasonable than the special case. Thanks!

I'm a bit worried that something else topples over because the very first timeout will now not be 0.0, but we'll see what CI says.